### PR TITLE
Dark theme

### LINKED ISSUE #43: Dark theme
Please save our eyes. And batteries. Do not ignore this property of useragent https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme

Cross-defect with https://bugs.python.org/issue39518

### PR DESCRIPTION
**Pull request by septatrix**
Closes #43. 
Closes https://github.com/python/python-docs-theme/issues/101. 

# Update: I now consider this ready

I have made a crude dark them as requested in #43, however there are still many unfinished areas.
So far I've always tested based on the `datetime.rst` file from the official docs as an example, though this does not use all styles which need to be changed.

# TODO

### General areas

- [x] Implement basic dark theme functionality
- [x] Implement proper dark mode syntax highlighting (merged for Sphinx 3.0)
- [x] Check complete docs for missing styles (e.g. find where `.refcount` and `.stableabi` are used)
- [x] Add functionality to toggle theme (persistence would require JS or server side code)
  - [x] Maybe add small disclaimer on first visit to show how to change to dark theme (decided not to as it is just another annoying banner like cookies on other websites)
- [x] Add dark mode compatible variants of images (e.g. pathlib images - maybe open issue in cpython repo) or alternatively add a white background for now (out of scope and has to be done in the python repo)

### Components / Classes
- [x] Check contrast ratios
  - [x] Fix styles for highlights after a search (kinda, feedback welcome; too bright makes text unreadable, too dark makes highlights hard to spot; went for a halfway point but maybe an outline is better)
- [x] Scrollbars (https://github.com/python/python-docs-theme/pull/44#issuecomment-584425632)
- [x] Sidebar collapse panel

### Design decisions

I choose the current main background color based on the header on [python.org](https://www.python.org/) and went for a different one (`#121212` from material design guidelines) for the background color around the edges as I thought this added a bit more depth to it. On second thought however it may be better to go with the single color approach as it is in the official docs. If we then were to the use the darker one this would also improve contrast and leave more space for fine tuning e.g. the anchor/link color.
Most other colors are just from experimenting with the color wheel and by what felt right. The background for syntax however still needs to change but for that topic also see below.

### Syntax highlighting

As neither Sphinx nor Pygments support switching themes based on a media query right now this would need to be a bit hacky. I might create an issue there and ask if they would be fine with pull request implementing such functionality (in whatever way though I already have an idea). Otherwise we would need to include another stylesheet after `pygments.css` overwriting the normal one.

### Theme toggling

As not everybody likes to always have a dark theme it may be a good idea to implement theme switching functionality. This could be achieved using either CSS and some label/input magic or plain ole JS (or server side). CSS would be very accessible but not persistent. Therefore I think we should use the CSS way but save the state in a cookie (like we already do for whether the sidebar is expanded or not).

# Try it out

## Live version
Updated somewhat regularly:
https://septatrix.github.io/cpython-dark-docs/

## Screenshot (outdated)

<details>
<summary>Full size screenshot - rather large</summary>

0 0 0 0_8000_html_datetime html
</details>

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept this contribution by verifying everyone involved has signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

## Recognized GitHub username

We couldn't find a [bugs.python.org](https://bugs.python.org/) (b.p.o) account corresponding to the following GitHub usernames:

@septatrix

This might be simply due to a missing "GitHub Name" entry in one's b.p.o account settings. This is necessary for legal reasons before we can look at this contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.




You can [check yourself](https://check-python-cla.herokuapp.com/) to see if the CLA has been received.

Thanks again for the contribution, we look forward to reviewing it!

**Comment by vitaly-zdanevich** [comment]
Thank you for your important work!

I would like to make background a little bit darker, with less contrast of text.

File: python_docs_theme/static/pydoctheme-default.css
Comment (vitaly-zdanevich) [inline python_docs_theme/static/pydoctheme-default.css]: Maybe always prefer shorter #eee for smaller CSS?

**Comment by vitaly-zdanevich** [comment]
Dark scrollbar?

**Comment by vitaly-zdanevich** [comment]
Just idea - maybe like in Reddit for Android - it is good idea to have Dark theme and AMOLED dark, with #000 as a background?

**Comment by JulienPalard** [comment]
Thanks for the work on this, I'm bad at UI/UX so I won't say anything more, I just appreciate someone is handling this.

**Comment by opensource-assist** [comment]
It's already very awesome. I like this colour.

**Comment by septatrix** [comment]
Thanks for the feedback so far. I will try to incorporate it as far as I can next week but before I finalize it I would like to await the response for dark theme support in Sphinx first.

File: python_docs_theme/static/pydoctheme-default.css
Comment (septatrix) [reply python_docs_theme/static/pydoctheme-default.css]:
```suggestion
    background-color: #eee;
```

Yes switching to the shorthand may be better though this was just the way it is in the current stylesheet...

**Comment by septatrix** [comment]
> Dark scrollbar?

Styling of scrollbars is not easy across browser vendors though I think we may use [`::-webkit-scrollbar`](https://developer.mozilla.org/en-US/docs/Web/CSS/::-webkit-scrollbar) for Chrome + Safari (+ new Edge) and [`scrollbar-color`](https://developer.mozilla.org/en-US/docs/Web/CSS/scrollbar-color) (still a working draft) for Firefox.

I currently have a few exams coming up so I probably won't get to implementing it until this weekend

**Comment by septatrix** [comment]
I did some further improvements and updated the TODO list above...
Here is what I've got so far, feel free to check it out:
https://septatrix.github.io/cpython-dark-docs/ (Docs from cpython repo)

**Comment by zacps** [comment]
Seems to be missing styles for 'sidebars'. See [asyncio](https://septatrix.github.io/cpython-dark-docs/library/asyncio.html#module-asyncio).

**Comment by septatrix** [comment]
I am currently working on finalizing the upstream changes. Additionally I have implemented a persistent disabling of the dark theme however no activation of the dark theme when prefers-color-scheme is light as this would be a bit trickier and ugly.
I will try to update the demo page today.

**Comment by septatrix** [comment]
### Updated the demo page

**Comment by vitaly-zdanevich** [comment]
I just want to say big THANK YOU again for continuing adding dark theme, this is important for many users.

**Comment by septatrix** [comment]
Thanks :D It is almost finished now. All that is left to do is think about how to handle images and maybe add an option to the theme to completely disable dark mode in the config. The upstream changes in Sphinx have been merged for 3.x so we should be good merging this next month when Sphinx 3 is released...

**Comment by septatrix** [comment]
As the Sphinx 3 release is right around the corner I did some fine tuning (e.g. reducing bright contrast to reduce eye strain) and switched to `localStorage` instead of `document.cookies`. Finally I implemented implemented a `forceDark` mode so the dark theme is also available to users without `(prefers-color-scheme: dark)` so there are now `forceLight`, `forceDark`, and `auto` (default)

If someone would be so kind to make a last review I would be very happy - I updated the live demo and otherwise would say it is ready to be merged :D

**Comment by vitaly-zdanevich** [comment]
Again - thank you for your work, this is important.

Checked demo right now: I want darker background than this:
image

**Comment by septatrix** [comment]
Do you have a color in mind? I was using material design, some code editor themes and a few websites as reference. Most of them were close to the one I choose or at least did not use very dark backgrounds. However using the current mechanism it would be very easy to implement an additional full black AMOLED theme if that is what you want...

**Comment by vitaly-zdanevich** [comment]
Yes, I want darker background, and I believe many other guys like me :)

In some apps they have not only light or dark themes, but also AMOLED dark, for example Reddit on Android, but I am not sure if we need three states. Maybe #111?

image

Also I think that vertical line at the left is too bright.

**Comment by septatrix** [comment]
I dimmed the color a bit and added an indicator what the current theme is in the button.
Regarding the colors I now went for `#222` for the bg though I personally am not willing to go any darker. If you check e.g. editor themes most of them are in the range of `#222` to `#333` as anything below that is very tiring for the eyes after a longer time

**Comment by vitaly-zdanevich** [comment]
ok, anyway, thank you again for such important work.

**Comment by septatrix** [comment]
Is there anything you want to be resolved before merging? Otherwise I would be pleased if you would merge this. In case you still want to adjust some colors I am certain you can do so on your own.

**Comment by septatrix** [comment]
Any update on whether this will get merged @vitaly-zdanevich

**Comment by aeros** [comment]
@theacodes would you care to take a look over this when you find the time? I think a dark theme for docs.python.org would be great and this looks decent at a surface level, but I don't have much proficiency w/ CSS.

**Comment by vitaly-zdanevich** [comment]
If not making the background a liiiitle bit darker, I have nothing to say, anyway, this is a huge step to the dark ecosystem :)

**Comment by septatrix** [comment]
@theacodes any feedback or estimate when you will be able to take a look at this?

**Comment by lysnikolaou** [comment]
Ζζ

**Comment by ParkHost** [comment]
@theacodes 
# The docs needs a dark mode, am blinding myself

**Comment by theacodes** [comment]
Hi friends. I'm currently on mental health leave. Please file a bug at
bugs.python.org and one of the other maintainers should be able to take a
look.

**Comment by septatrix** [comment]
This should probably be squashed before merging. Will this be performed by you or should I do this manually?

**Comment by JulienPalard** [comment]
I'm not a front-end guy, but I think the PR would be more readable without reindenting the whole CSS file, can you keep a 4 spaces indentation to avoid this huge diff?

(Having a smaller diff would also help avoid conflicts with others PRs modifying the CSS where you did not).

**Comment by JulienPalard** [comment]
I tried a local build using [docsbuild-scripts](https://github.com/python/docsbuild-scripts/):

    $ rm -fr build_root/venv-with-sphinx-2.4.4/  # Because it won't force-reinstall the theme
    $ python3 ./build_docs.py --quick --build-root ./build_root --www-root ./www --log-directory ./logs --group $(id -g) --skip-cache-invalidation --theme git+https://github.com/septatrix/python-docs-theme@dark-theme --branch 3.9 --language en

I'm using Firefox 86.0.1 with the Dark theme, and it's ... strange:

Screenshot_2021-03-24 3 9 2 Documentation

**Comment by JulienPalard** [comment]
I just noticed the "change theme" button, why not just relying on the browser configuration?

**Comment by septatrix** [comment]
> I'm not a front-end guy, but I think the PR would be more readable without reindenting the whole CSS file, can you keep a 4 spaces indentation to avoid this huge diff?

Yes sure this was just my IDE.

> > This should probably be squashed before merging. Will this be performed by you or should I do this manually?
> 
> Totally agree :P we do have "squash and merge" enabled on this repo, and if you ask so, it should be used, yes. This is the one I would use to merge such a PR.

Okay, then I will not bother with that.

> I'm using Firefox 86.0.1 with the Dark theme, and it's ... strange

Hm I tried this with Firefox locally and did not see this behaviour. I will look into this again.

> I just noticed the "change theme" button, why not just relying on the browser configuration?

The default theme is "auto" which respects the `prefers-color-scheme` setting. However some users would probably like to override this which is why I added the button which cycles between auto, dark and light.

**Comment by pradyunsg** [comment]
Hiya!

Could this PR be rebased and https://septatrix.github.io/cpython-dark-docs/ updated?

There's also fairly sweeping formatting changes in this PR that make it difficult to see what exactly has changed in the CSS file. Could those changes be omitted or made in a separate PR?

Also, one nit pick I have is using the text "toggle color scheme" on the theme toggle instead of "toggle theme".

**Comment by JulienPalard** [comment]
> and https://septatrix.github.io/cpython-dark-docs/ updated?

I don't know if it's a lot of work to update it, but any way there's a build artifact now (produced if you rebase on top of master) so anyone can test locally.

> Also, one nit pick I have is using the text "toggle color scheme" on the theme toggle instead of "toggle theme".

I'm personally not fan of the "toggle" button, I made it tested by multiple persons and the reports were "it does not work", because of the one every three toggle doing nothing. I'd prefer a first merge without a button at all (only browser opt-in) to take the time to find a good solution for a button toggler (if it's really needed, as I use my browser to switch themes, I personally don't feel the need of a button, I'm biased).

**Comment by JulienPalard** [comment]
For the reviewers: in the "Checks" tab, there's one artifact that can be downloaded, unzipped, and tested locally with `python3 -m http.server`.

**Comment by septatrix** [comment]
> [...] but distribs are packaging the Python documentation too, and they may not have Sphinx 3, [we have to keep compatibility with Sphinx 1.8](https://github.com/python/cpython/blob/d1b81574edd75e33ae85c525ac988ce772675a07/Doc/conf.py#L48)

Hm that is a bummer. Do you know if there is a possibility in sphinx to create a different behaviour based on the version (it seems there is a global `sphinx_version` which may be what we want)? While most parts of a dark mode could still be implemented code blocks would be highlighted using the default light stylesheet. The possible solutions I see are a) live with it and let docs created with old sphinx version have light syntax highlighting, b) if possible invert the colors of code block if an old sphinx version is detected or c) raise the version requirement for sphinx.

The last one does not seem to far fetched as distributions shipping the docs are free to use an older version of the theme and recent releases of most major distributions (e.g. Ubuntu, Fedora, upcoming Debian Bullseye) are already shipping new sphinx versions.

**Comment by septatrix** [comment]
This is still blocked on using Sphinx 3. Without that version it is not possible to use two pygments styles at the same time. I consider having light code with a dark documentation or vice versa as a big enough nuisance that without I would not implement a dark mode without that working.

**Comment by JulienPalard** [comment]
We still have to keep compat with Sphinx 1.8, so forgot about bumping to requireing Sphinx 3 in a near future.

The workaround that could be used is: detect Sphinx version, disable dark theme on Sphinx < 3. As we have Sphinx > 3 in docs.python.org it's good enough.

**Comment by pradyunsg** [comment]
Can't we use an older version of this theme on older Python versions?

**Comment by JulienPalard** [comment]
> Can't we use an older version of this theme on older Python versions?

I don't think we can use [PEP 508](https://www.python.org/dev/peps/pep-0508/) conditional markers to make the switch: if some distribs still use Sphinx 1.8, I don't want to try to guess which version of pip they have.

We can still use a specific version for docs.python.org and pin an old one in our Makefile though, it's not nice, but it'll work, yes.

**Comment by elsiehupp** [comment]
If you'd like to simultaneously squash and rebase your commits, a simple way of doing so is as follows:

```bash
$ git reset --soft master
$ git commit -s -m "<commit_message>"
$ git push <origin_or_upstream> +<existing_branch>
```
where "<origin_or_upstream>" refers the name your local `git` uses for the `septatrix/python-docs-theme` remote (usually `origin`) and "<existing_branch>" is the existing branch name, e.g. `dark-theme`.

Also I included the option `-s`, which adds a [DCO](https://developercertificate.org/) signature to the commit message, if that's a thing you care about (or a particular project requires). You can probably remove it here.

With the placeholder values replaced, it would probably look something like this:

```bash
$ git reset --soft master
$ git commit -m "Add dark theme"
$ git push origin +dark-theme
```

**Comment by elsiehupp** [comment]
If you'd like to make your changes clearer in the record, it might be useful to segment them into two sequential commits or even two separate pull requests:

1. Extract/separate the existing color assignments from the main CSS, while maintaining existing functionality.
2. Add the dark-theme color assignments and the logic for enabling/disabling them.

Depending on how you structure things, one way of doing a safe fallback is to use `!important` throughout the dark-mode CSS in order to override the light-mode values, rather than disabling them. This behavior would make auto dark mode fail gracefully on browsers that don't support it, but it would be less necessary if you include the toggle.

Another benefit to the `!important` approach is that makes it unnecessary to extract the color values from the existing CSS, which reduces the overall number of lines changed. This way you could instead have two separate pull requests as follows:

1. Add the `!important` override dark-theme color assignments and the logic for enabling/disabling them.
2. Optionally, later on, make a second pull request refactoring the structure of the CSS if the changes in the first pull request feel brittle or kludgy.

Which of the two approaches is better basically just depends on what the project maintainers prefer. (I'm only making these suggestions because it seems like the maintainers have been sitting on this pull request, so there might be something they're uncomfortable with, idk.)

**Comment by NewUserHa** [comment]
I think the current https://septatrix.github.io/cpython-dark-docs/ is too much on contrast and will damage your eyes.
The dark theme color should simply learn the `vscode` `default dark` or the `Windows` or the `youtube`, or even any terminal color.

BTW, the first demo in this thread looks like some copper solution..

**Comment by JulienPalard** [comment]
> will damage your eyes

I hope you're exaggerating (if not, please provide source showing it can cause more than fatigue).

**Comment by NewUserHa** [comment]
'damage' -> 'not comfortable'.
I only see high contrast theme on windows or TVs or macos(?), so some good normal dark theme can probably work, and the ppl can can use system settings for the high contrast needs.
for the color, because ppl with different monitor will see different color. high-end monitor see different than the economic monitors, so use the exists color from those good user experience already products things like: vscode, windows, or youtube can be better for python document users.

**Comment by septatrix** [comment]
> [...] so use the exists color from those good user experience/well-designed already products things like: vscode, windows, or youtube can be better for python document users, because they can use very various monitors.

VSCode Dark is #1e1e1e, youtube is #181818, windows I do not know, Chrome DevTools is #212124. Some of these like yt still use pure white which we also do not do so I think the color right now are quite well.

Nevertheless this is still blocked by not being able to use sphinx 3 which is necessary to also switch the color theme of code blocks.

**Comment by NewUserHa** [comment]
your color #222 in your site, is higher/brighter than all other products like VSCode Dark, youtube, windows, etc. and the font color in your site is very brighter too.
on my monitor(a good designer one), it looks shaper but not clear. if on OLED monitor, I guess it's the similar result(not clear). why use your favorite color on your monitor than the well-designed from other world-wide good product?(no offense)
btw, what your monitor model is?

**Comment by NewUserHa** [comment]
https://doc.rust-lang.org/book/ documents for rust lang has 5 themes(Coal looks better on my monitor than others). can python have multi theme choice too?
vscode dark, youtube, windows dark themes probably are not bad choices.

**Comment by septatrix** [comment]
> I think the current https://septatrix.github.io/cpython-dark-docs/ is too much on contrast

> your color #222 in your site, is higher/brighter

There is a difference between a contrast which is too high and the color being too bright.

> Can python have multi theme choice too?

With how it currently works it would be possible to implement though not for the embedded code blocks which are limited to a single dark theme.

**Comment by jdevries3133** [comment]
Hi everybody, so I am kind of thick and started working on a dark theme implementation without taking 2 seconds to read the open PRs and issues (oops -_- ).

You can look here to see exactly what I've done so far:

https://github.com/python/python-docs-theme/compare/master...jdevries3133:feat/dark-mode?expand=1

Anyway, I started by factoring all the colors into css variables, which seemed to clean up the CSS quite a bit, allow me to remove many redundant and repetitive colors, and probably opens the door to more theming in the future. Clearly, there's been a lot of work on this PR, and I literally *just* factored out CSS colors into variables (no work towards the actual dark color scheme).

However, I do have criticisms of this implementation. Adding a separate stylesheet for dark mode is very non-DRY, and it only gets worse if you eventually want dark mode low contrast, dark mode high contrast, etc. – which I saw discussed already in this PR thread (https://github.com/python/python-docs-theme/pull/44#issuecomment-609105027). Moving to CSS variables, on the other hand, is going to allow for a wide variety of theming without much repetition. I'm interested in what @septatrix thinks about CSS variables versus a separate stylesheet. @septatrix, since you were the first one to take on this issue, I'm more than happy to step aside while you run with the CSS variable suggestion, or I'd be enthusiastic about taking a stab at this feature if I have your blessing. I certainly don't want to step on your toes.

**Comment by septatrix** [comment]
> However, I do have criticisms of this implementation. Adding a separate stylesheet for dark mode is very non-DRY, and it only gets worse if you eventually want dark mode low contrast, dark mode high contrast, etc. – which I saw discussed already in this PR thread ([#44 (comment)](https://github.com/python/python-docs-theme/pull/44#issuecomment-609105027)).
> Moving to CSS variables, on the other hand, is going to allow for a wide variety of theming without much repetition. I'm interested in what @septatrix thinks about CSS variables versus a separate stylesheet. 

This was done for a few reasons:

* IE 11 which does not support CSS vars was more prevalent 2 years ago though that should not be a concern any longer,
* the amount of code repetition was very low and I did not deem refactoring it worth it,
* and lastly this it the way the dark-theme support in Sphinx works (for pygments, i.e. syntax highlighting).

> @septatrix, since you were the first one to take on this issue, I'm more than happy to step aside while you run with the CSS variable suggestion, or I'd be enthusiastic about taking a stab at this feature if I have your blessing. I certainly don't want to step on your toes.

I am fine either way.

**Comment by AA-Turner** [comment]
Perhaps of interest, the PEPs theme uses CSS variables and we've had no complaints.

https://github.com/python/peps/blob/4b04a2b7bed7b88a97d5830c7bcead71de89f80f/pep_sphinx_extensions/pep_theme/static/style.css#L33-L43

Sphinx is also dropping support for any browser that doesn't support ES2018 in Sphinx 5, so that may serve as a lower bound for the compatability matrix.

A

**Comment by AA-Turner** [comment]
> We do use Sphinx v3 in production for 3.10 and 3.11 on docs.python.org, but distribs are packaging the Python documentation too, and they may not have Sphinx 3, [we have to keep compatibility with Sphinx 1.8](https://github.com/python/cpython/blob/d1b81574edd75e33ae85c525ac988ce772675a07/Doc/conf.py#L48)

@JulienPalard is there a list of distributions that don't have a modern Sphinx? (Or a tracker issue for the Sphinx version discussion that I could subscribe to please?)

I had missed that comment originally -- I still think it is reasonable to require CSS variables, ES2018, but I'll defer to those with much more experience.

A

**Comment by JulienPalard** [comment]
> @JulienPalard is there a list of distributions that don't have a modern Sphinx? (Or a tracker issue for the Sphinx version discussion that I could subscribe to please?)

You can subscribe to https://github.com/python/cpython/pull/92318, it's currently being discussed.

File: .github/workflows/tests.yml
Comment (septatrix) [reply .github/workflows/tests.yml:32]: This was added by Julien while the workflow was broken

**Comment by septatrix** [comment]
Sorry to leave you in the dark for so long (heh) but I was now able to take a bit of time and rebase it to work atop master. I had to make a few small changes, especially to the hover/focus styles, of the mobile navbar but I think it look good now while in dark mode. (Also the scrollbars are properly updating switching to dark mode now, too.)

As usual, I updated the snapshot: <https://septatrix.github.io/cpython-dark-docs/>

**Comment by NewUserHa** [comment]
The background color looks fine on my devices.

But the font color looks not good. fonts shouldn't have any alpha channel.

**Comment by nk9** [comment]
I have done a spot check for WCAG compliance with the dark theme. Links in inline code text [is not WCAG compliant](https://accessibleweb.com/color-contrast-checker/) for color contrast.

image

Nor is comment text in code blocks:

image

I'm on mobile right now, so there is probably more to review. But I wanted to highlight these two issues.

**Comment by septatrix** [comment]
> But the font color looks not good. The fonts shouldn't have any alpha channel.

I will take another look at this. The alpha channels can probably be removed and the color just be a fixed slightly grayish white (to be easier on the eyes).

> I have done a spot check for WCAG compliance with the dark theme. Links in inline code text [is not WCAG compliant](https://accessibleweb.com/color-contrast-checker/) for color contrast.

Do you have an example of that? For an `a > code > span.pre` element the devtools show me a color of `#77aaff` on top of `#222222` which results in a contrast of 6.8. Regardless, there definitely are a few areas with undesirable contrast ratios, especially some of the boxes (info, warning, aside) and the code highlighting. For those, however, I think that adjusting those would outweigh the benefits as the alternative would be e.g. a high contrast syntax highlighting theme which often look rather bad and don't provide nearly as well of a job of highlighting the syntax. If you were to check the existing light theme you will discover that it too does not have ideal contrast ratios everywhere.

For areas where the contrast is abysmal I will try to improve them in this PR, otherwise I think separate follow-up issues are better (and could then also target the light theme).

**Comment by hugovk** [comment]
Tip: Chrome has tools for finding and suggesting fixes for low contrast.

1. Go to https://septatrix.github.io/cpython-dark-docs/whatsnew/3.11.html
2. Open Chrome devtools and the Lighthouse tab (may need enabling in settings)
3. Run an accessibility audit (can disable the others for speed)
4. That will show a number of code blocks with low contrast, for example:

image

5. Inspect the element and find the `color` style:

image

6. Click the colour box to see the contrast ratio, and AA and AAA guidelines with suggestions:

image

7. Click the colour suggestions next to AA or AAA to apply that colour:

image

AA is a good minimum target. AAA can sometimes be a bit too far away from the desired colour, but it's worth considering.

But in this case, we're using Pygments for _code block_ styling, so these reports/fixes should rather go upstream to https://github.com/pygments/pygments so everyone benefits and we're not overriding future fixes. For example, I've already contributed some fixes: https://github.com/pygments/pygments/pull/2038.

---

https://wave.webaim.org/report#/https://septatrix.github.io/cpython-dark-docs/whatsnew/3.11.html is another good tool. From a quick scan, I only see contrast issues in code blocks.

**Comment by hugovk** [comment]
CI build error will be fixed by PR https://github.com/python/python-docs-theme/pull/94.
```
  File "./docsbuild-scripts/build_docs.py", line 46, in <module>
    import requests
ModuleNotFoundError: No module named 'requests'
```

**Comment by septatrix** [comment]
> Tip: Chrome has tools for finding and suggesting fixes for low contrast.

Yes, there is also the "CSS Overview" tab which lists contrast issues though the Lighthouse tool is more complete regarding general accessibility.

> But in this case, we're using Pygments for code block styling, so these reports/fixes should rather go upstream [...]

Definitely. Though some themes like Monokai (the one I selected here) is not created by the Pygments authors and is an exact copy of a preexisting popular theme so I do not think it will get changed. We could, however, select a different dark theme which might provide better contrast rations.

**Comment by nsandary** [comment]
Any update here? Python Docs is one of the last places of eye-searing pain left in my daily workflow...

**Comment by septatrix** [comment]
> Any update here? Python Docs is one of the last places of eye-searing pain left in my daily workflow...

From my side this is ready, especially the technical side. Fine tuning some colour tones can still be done later on, as can another theme like "True black" (for AMOLED screens) be implemented. Everyone who wants to try this out is free do to so and maybe check for significant contrast issues or rare elements which I might have missed during the conversion and are still light.

Due to the reasons stated above, however, I do not think that we should focus on contrast issues within Pygments, as those are out of scope and high contrast code highlighting always looks ugly (and would be another potential suspect for a separate theme - which again, can be done in follow up PRs).

**Comment by hugovk** [comment]
I think this is looking really good!

I agree minor things can be dealt with later, and if we don't find any really major things it would be good to merge and iterate from there.

And I agree Pygments "Monokai" code contrast issues don't really deserve a fix here, they should usually be reported/fixed upstream at https://github.com/pygments/pygments (I've already done a few things for the "Native" dark style used by https://github.com/python/peps).

File: python_docs_theme/layout.html
Comment (hugovk) [inline python_docs_theme/layout.html:100]: Is this empty `<span>` needed?
```suggestion
```

**Comment by hugovk** [comment]
On my phone (Android, Samsung S10, Chrome), the search box expands off the screen, which the current docs don't do:

<details>
<summary>Screenshots</summary>

<table>
<tr>
<td>
<td>
</table>

</details>

This is partly an existing problem in that the whole page can be scrolled to the right (new theme and existing theme), but now the search box now extend to the off-screen section:

<details>
<summary>Screenshots</summary>

<table>
<tr>
<td>
<td>
</table>

</details>

**Comment by nmstoker** [comment]
I'm late to hearing about this but it's good to see this being worked on.

Is the plan to get this out and then refine the design / colour options?

I'm viewing it on Firefox Nightly on Android (I'll try to compare with other browsers/computers later on in case there are significant differences).

Right now the colour choices, especially around the various shades of grey, seem quite jarring. I'm no colour expert but I suspect it might look more consistent if most of the background shades came from essentially the same "colour" / hue of grey and largely varied the lightness (subtly). Also having shades of grey and then a pure black header looks odd too.

**Comment by pdcastro** [comment]
My 2 cents is that I have been using the dark theme [preview](https://septatrix.github.io/cpython-dark-docs/) for a while instead of the of official docs, and I find that the dark theme is a massive improvement _*just as it is right now.*_ This PR was created 3 years ago and the worse outcome would be for the recent momentum to be lost and the PR to be abandoned. So I would say:

* Fix/change only what is absolutely necessary, i.e. issues that prevent the site from being used. I am not aware of any such issue.
* Get it merged and released.
* Leave other minor issues for future PRs, e.g.:
  - The search box expanding off the screen on mobile.
  - WCAG compliance on color contrast.
  - Comments on color shades and alpha channels.

As soon as this PR is merged and released, invite everyone who commented here on the remaining unaddressed issues to create new corresponding GitHub issues, to be fixed in future PRs by whoever is willing to contribute further.

This is not to dismiss the remaining issues as unimportant. On the contrary, getting this PR merged increases the chance of the remaining issues being addressed because they become doable as smaller pieces of work, by a wider pool of contributors.

Congrats to everyone who put effort creating this PR, contributing and reviewing, especially @septatrix - great job! 👏 🎉

File: python_docs_theme/layout.html
Comment (septatrix) [reply python_docs_theme/layout.html:114]: I noticed a discrepancy between the html fragment for the desktop search form and the one shown in the mobile view. However, they do not seem to fulfil any purpose nowadays: https://github.com/sphinx-doc/sphinx/pull/5511

File: python_docs_theme/layout.html
Comment (septatrix) [reply python_docs_theme/layout.html:100]: Yes, it is parent to before and after pseudo-elements which form the hamburger/X icon

**Comment by septatrix** [comment]
> On my phone (Android, Samsung S10, Chrome), the search box expands off the screen, which the current docs don't do

Fixed

**Review by hugovk — APPROVED** [review]
Current preview link: https://python-docs-theme-previews--44.org.readthedocs.build/en/44/

---

I asked people on [Mastodon](https://mastodon.social/@hugovk/109824258264361734) to try it out and only got positive feedback.

I agree this is good to merge now, and other things can be adjusted in followups. 

Thank you!

**Comment by pradyunsg** [comment]
With #109, I think we can click merge on this?

**Comment by hugovk** [comment]
Almost! Let's deal with https://github.com/python/python-docs-theme/pull/108#issuecomment-1464879086 first.

**Comment by joshhacksthings** [comment]
@hugovk now that the [#108 (comment)](https://github.com/python/python-docs-theme/pull/108#issuecomment-1464879086) has been fixed with #113 , can this now be merged? 😀

**Comment by hugovk** [comment]
Yes, let's do it.

@septatrix Thank you for all your work on this PR! And to everyone else for the feedback.

**Comment by JulienPalard** [comment]
Just ran:

    docsbuild@docs:/srv/docsbuild$ ./venv-3.12/bin/python -m pip install git+https://github.com/python/python-docs-theme/

manually so maybe in a few hours docs.python.org/dev/ will use the dark theme, I think it's a good way to test it.

**Comment by hugovk** [comment]
See also https://github.com/python/cpython/issues/103960. Not a theme issue, we need to adjust some images in CPython docs for dark mode.

**Comment by hugovk** [comment]
Please see PR https://github.com/python/python-docs-theme/pull/128 to add the class to the theme.

**Comment by brutatoasta** [comment]
It's live on python 3.13 docs, but not on previous doc versions as recent as python 3.11. Any chance to push the changes there too?

**Comment by hugovk** [comment]
This is now live at the main https://docs.python.org site (and [3.12](https://docs.python.org/3.12/) and [3.13](https://docs.python.org/3.13/)), thanks again @septatrix for all your work and patience!

image

**Pull request merged**
